### PR TITLE
feat: Support GatewayAPI route attachment restriction

### PR DESCRIPTION
**Pull request by stillfox-lee**
### Type of change:

close: https://github.com/apache/apisix-ingress-controller/issues/1424
close: https://github.com/apache/apisix-ingress-controller/issues/1349

### What this PR does / why we need it:

- Add new feature for support GatewayAPI route attachment
- Fix https://github.com/apache/apisix-ingress-controller/issues/1424. `GatewayClass` is cluster scoped resource, shouldn't  filter by namespace.
- Fix `Gateway` sync function. Compatible with scenario: `GatewayClass` add event is received after the `Gateway` event.

### CODE REVIEW

File: pkg/providers/utils/string.go [hunk 1]
```diff
@@ -45,3 +45,11 @@ func Equal(a, b []string) bool {
 	}
 	return len(Difference(a, b)) == 0 && len(Difference(b, a)) == 0
 }
+
+func ReverseString(s string) string {
+	var reversed string
+	for _, v := range s {
+		reversed = string(v) + reversed
```
Comment (lingsamuel) [inline pkg/providers/utils/string.go:52]: use rune array and convert it to string in return statement would be better than convert rune to string every iteration.

File: pkg/providers/gateway/provider.go [hunk 2]
```diff
@@ -274,6 +286,30 @@ func (p *Provider) RemoveListeners(ns, name string) error {
 }
 
 func (p *Provider) FindListener(ns, name, sectionName string) (*types.ListenerConf, error) {
+	p.listenersLock.Lock()
```
Comment (lingsamuel) [inline pkg/providers/gateway/provider.go:289]: RLock is enough here

File: pkg/providers/gateway/provider.go [hunk 3]
```diff
@@ -274,6 +286,30 @@ func (p *Provider) RemoveListeners(ns, name string) error {
...
+	for _, listener := range listeners {
+		if listener.SectionName == sectionName {
+			return listener, nil
+		}
+	}
 	return nil, nil
 }
+
+func (p *Provider) QueryListeners(ns, name string) (map[string]*types.ListenerConf, error) {
+	p.listenersLock.Lock()
```
Comment (lingsamuel) [inline pkg/providers/gateway/provider.go:306]: RLock is enough here

File: pkg/providers/gateway/validator.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+	case *gatewayv1alpha2.UDPRoute:
+		r.routeNamespace = route.Namespace
+		r.parentRefs = route.Spec.ParentRefs
+		r.routeProtocol = gatewayv1alpha2.UDPProtocolType
+		r.routeGroupKind = gatewayv1alpha2.RouteGroupKind{
+			Group: &group,
+			Kind:  types.KindUDPRoute,
+		}
+	default:
+		return nil, fmt.Errorf("validator unsupported Route")
```
Comment (lingsamuel) [inline pkg/providers/gateway/validator.go:101]: Would be better to include the exact type in the error message

File: test/e2e/scaffold/ingress.go [hunk 5]
```diff
@@ -200,6 +200,14 @@ rules:
...
     - list
     - watch
+  - apiGroups:
+    - gateway.networking.k8s.io
+    resources:
+    - gateways/status
+    - gatewayclasses/status
+    verbs:
+    - get
+    - update
```
Comment (lingsamuel) [inline test/e2e/scaffold/ingress.go:210]: We can also add `list` verb here

File: test/e2e/suite-gateway/route_attchment.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  listeners:
+  - protocol: HTTP
+    port: 80
+    name: same-namespace-route
+    allowedRoutes:
+      namespaces:
+        from: Same
+  - protocol: HTTP
+    port: 80
+    name: crocss-namespace-route
```
Comment (lingsamuel) [inline test/e2e/suite-gateway/route_attchment.go:61]:
```suggestion
    name: cross-namespace-route
```

Comment (lingsamuel) [reply test/e2e/suite-gateway/route_attchment.go:61]: same typos below

File: pkg/providers/utils/domain.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			return idx == lLength-1
+		}
+		if labels == &rLabels {
+			return idx == rLength-1
+		}
+		return false
+	}
+
+	// reverse matching, hostname seq like ["moc", "xisipa", "ssergni"]
+	for i := 0; i < lLength || i < rLength; i++ {
```
Comment (lingsamuel) [inline pkg/providers/utils/domain.go:52]: Using two variables (`i,j`) here can make the logic clearer and simpler.

**Comment by stillfox-lee** [comment]
Hi @tao12345666333 @lingsamuel @AlinsRan .
Since https://github.com/apache/apisix-ingress-controller/pull/1445, `Gateway` and `HTTPRoute` has been upgraded to v1beta1. But other route resources are still under v1alpha2. Hence I need to implement both v1alpha2 and v1beta1 validation to support all route resources. If we planning to support multiple versions. Why `Translator` just replace the older version? I'm confused about our GatewayAPI's versioning management.

**Comment by AlinsRan** [comment]
> Hi @tao12345666333 @lingsamuel @AlinsRan . Since #1445, `Gateway` and `HTTPRoute` has been upgraded to v1beta1. But other route resources are still under v1alpha2. Hence I need to implement both v1alpha2 and v1beta1 validation to support all route resources. If we planning to support multiple versions. Why `Translator` just replace the older version? I'm confused about our GatewayAPI's versioning management.

TCPRoute、UDPRoute etc. have not entered v1beta1. 
In fact, only one version needs to be implemented, and gateway-api CRDs will be compatible with it. The final effect is that only one stable version needs to be maintained.

**Comment by tao12345666333** [comment]
as @AlinsRan said
we only need maintain only one version.

**Comment by stillfox-lee** [comment]
> > Hi @tao12345666333 @lingsamuel @AlinsRan . Since #1445, `Gateway` and `HTTPRoute` has been upgraded to v1beta1. But other route resources are still under v1alpha2. Hence I need to implement both v1alpha2 and v1beta1 validation to support all route resources. If we planning to support multiple versions. Why `Translator` just replace the older version? I'm confused about our GatewayAPI's versioning management.
> 
> TCPRoute、UDPRoute etc. have not entered v1beta1. In fact, only one version needs to be implemented, and gateway-api CRDs will be compatible with it. The final effect is that only one stable version needs to be maintained.

Thanks for your reply. I should RTFM at first.

**Comment by lingsamuel** [comment]
The code LGTM, just need to resolve the conflicts.
Thanks for your contribution!

**Comment by stillfox-lee** [comment]
@tao12345666333 I have run  the `golangci-lint` and no errors occurred. PTAL.

File: pkg/providers/gateway/types/types.go [hunk 8]
```diff
@@ -16,7 +16,17 @@
 // under the License.
 package types
 
-import gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+import (
+	"github.com/apache/apisix-ingress-controller/pkg/providers/utils"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
```
Comment (tao12345666333) [inline pkg/providers/gateway/types/types.go:21]: These imports need to be sorted

**Review by tao12345666333 — COMMENTED** [review]
LGTM,

Some imports require sorting, which is why lint fails.

Also, could you merge the master branch first?

**Comment by stillfox-lee** [comment]
ping @tao12345666333 @lingsamuel . All CI passed.

**Comment by tao12345666333** [comment]
sorry for delay, I recently got covid-19.

plase resolve conflicts, thanks!

**Comment by stillfox-lee** [comment]
@tao12345666333 All done. PTAL.

**Review by lingsamuel — APPROVED** [review]

**Review by AlinsRan — APPROVED** [review]

**Comment by tao12345666333** [comment]
Sorry for the delay, the code LGTM.

Please resolve the conflicts, thanks!

**Comment by tao12345666333** [comment]
The test case job failed. Please check it. Thanks

**Comment by stillfox-lee** [comment]
`2023-03-08T18:02:17+08:00	error	gateway/gateway.go:269	failed to record status change for Gateway resource	{"error": "Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"test-gateway\": StorageError: invalid object, Code: 4, Key: /registry/gateway.networking.k8s.io/gateways/ingress-apisix-e2e-tests-default-299106829/test-gateway, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 043033bc-3aa5-445c-b8bb-0149f5734dc2, UID in object meta: ", "name": "test-gateway", "namespace": "ingress-apisix-e2e-tests-default-299106829"}`
It appears from the information above that the UID of the Gateway resource has changed, making it unable to update the Status sub-resource, which has led to the test failing. I have tested this several times in my local environment, but have been unable to replicate the issue, so I may need more time to resolve this problem.

**Comment by stillfox-lee** [comment]
@tao12345666333 Please help me run the test again.

**Comment by AlinsRan** [comment]
I will rerun the failed job.

**Review by tao12345666333 — APPROVED** [review]

**Pull request merged**